### ISSUE TITLE
LSP: auto complete suggestions on pipes

### ISSUE THREAD

**Issue by gusinacio** [gleam-lang/gleam#3297]
I don't know how functional languages LSPs usually handle this case but it would be nice if suggestions after a pipe are functions that contain the type as the first parameter. i.e:
```
"hello" |> // <- This would only suggest functions that have String as the first parameter type
```

If there are multiple pipes, the suggestions should be related to the last output value.
```
"2" |> int.parse |>  // <- This would only suggest functions that have Result as the first parameter type
```

**Comment by lpil** [comment]
Good suggestion, thank you!

**Comment by LoipesMas** [comment]
I don't think it should _only_ suggest functions that have that type as the first argument, because you can use [function captures](https://tour.gleam.run/functions/function-captures/) in pipes.
So in case of `37 |> `  you may want to pipe it into `string.repeat("foo",_)`, which, normally, doesn't have `Int` as the first argument, so `string.repeat` wouldn't show up.
So it would have to suggest functions that have the type as any of the arguments.
Not sure about the LSP details, but maybe it's possible to first suggest functions that have the type as first argument, and then functions that have that type as any of the arguments?

**Comment by gusinacio** [comment]
This is interesting. I think it makes sense to at least filter out functions with no arguments at all and also functions that none of the arguments are of the pipe type.

After that it could suggest functions with the first argument as the pipe type, then the second argument and so on.

**Comment by lpil** [comment]
OK. I'm going to put this as low priority as it only makes a difference until the programmer types the first character, at which point that'll have more impact than any ordering on what is found, so it's work for very little gain.

**Comment by dancojocaru2000** [comment]
A somewhat related feature (not sure if I should make a new issue for it) would be to complete on dot in order to simulate OOP-style function calls. So `"2".` would suggest `int.parse`, and, when accepting the suggestion, it would become `int.parse("2")`.

**Comment by Olian04** [comment]
@dancojocaru2000 how would that handle cases such as:

```gleam

type Pet {
  Dog(name: String, age: Int)
  Cat(name: String, age: Int)
}

fn greet(pet: Pet) -> String {
  "I am " <> pet.name <> " and I am " <> int.to_string(pet.age) <> " years old"
}

fn unwrap_pet(pet: Pet) -> String {
  pet. // <---
}
```

What should be suggested here? Functions that use `Pet` as its only argument, or the properties available on all `Pet` variants, or both (in which case, which order should they show up in?)?

**Comment by dancojocaru2000** [comment]
Both, properties first? In my proposed complete-after-dot, only the first argument of functions should ever be considered.

In Python, methods are declared and used like this:
```py
class SomeClass:
    def some_method(self, param1):
        # do stuff here

some_var = SomeClass()
some_var.some_method("sample value")
```

`some_var` would be passed as the `self` parameter. My suggestion for completion takes that logic into account. 

In non-OOP languages, the "object" is also generally the first parameter of functions related to it. For example, `send(socket_fd, some_data, size_of_data, 0)` in C takes the socket file descriptor as the first parameter, and so do other functions that deal with sockets like `setsockopt`.

**Comment by D-matz** [comment]
poking this again, here a very clever LSP could tell I need a function that takes a DateTime and returns a Stringn but it puts a fn that takes Date first instead: ![Image](https://github.com/user-attachments/assets/504ed35f-75d2-4711-9629-7efc3737d15a) 

as @LoipesMas said "first suggest functions that have the type as first argument, and then functions that have that type as any of the arguments" imo the value is suggesting the correct fn for api discoverability

I am not 100% sure of all pipeline's valid uses, but if you sort (rather than filter) you wouldn't break anything, and pipeline to first arg of fn feels to me like the most conventional/least confusing use of |> so it'd be valuable in the sort of common 80/20 case

On the other hand this is not a super urgent pressing need and I might be missing some implementation challenges
